core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2098,11 +2098,55 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    val n         = as.size
+    val resultArr = new Array[Any](n)
+    val resultList: ZIO[R, E, List[B]] = for {
+      parentId <- ZIO.fiberId
+      latch    <- CountdownLatch.make(n)
+      failed   <- Promise.make[Nothing, Int]
+      cause    <- Ref.make[Cause[E]](Cause.empty)
+      fibers <- as
+                 .foldLeft[(Int, URIO[R, List[(Int, Fiber[E, B])]])]((0, URIO.succeed(Nil))) { (r, a) =>
+                   r match {
+                     case (i, acc) =>
+                       def handleExit(exit: Exit[E, B]): URIO[R, Any] = exit.foldM(
+                         c => cause.update(cs => cs && c) *> failed.succeed(i),
+                         b => ZIO.effectTotal(resultArr(i) = b)
+                       )
+
+                       val appended = for {
+                         fs <- acc
+                         f <- ZIO
```

Comment:
```scala
fn(a).foldCauseM(
  c => cause.update(cs => cs && c) *> failed.succeed(i),
  b => ZIO.effectTotal(resultArr(i) = b)
).ensuring(latch.countDown)
```

Reply:
That certainly looks nicer. Is it completely equivalent to what I wrote? (I'm not sure of the subtleties of the difference in semantics of  bracketExit and foldCauseM)

Reply:
Yes, in the sense the contract of `ZIO.bracket*(acquire, release, use)` is that `release` is guaranteed to be evaluated if `acquired` succeeded, but since `ZIO.succeed` cannot fail and `handleExit` is just `foldCauseM` in disguise, you can just remove the `bracketExit`.